#Rule that performs the alignments
rule align_reads:
	input:
    	fastq1 = "data/{sample}.1.fastq.gz",
		fastq2 = "data/{sample}.2.fastq.gz"
	output:
		bam = "results/{sample}.bam"
	resources:
		mem = 8000
	threads: 1
	shell:
		"""
		module load hisat-2.0.4
        module load samtools-1.6
		hisat2 -p {threads} -x annotations/hisat2_index/hisat2_index -1 {input.fastq1} -2 {input.fastq2} | samtools view -Sb > {output.bam}
		"""

#One meta rule whose input files are all 
#of the desired output files
rule make_all:
	input:
		expand("results/{sample}.bam", sample=config["samples"])
	output:
		"out.txt"
	shell:
		"echo 'Done!' > {output}"