---
title: "Association testing and fine mapping"
output: 
  html_document: 
    keep_md: true
date: "2025-04-06"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=FALSE, warning=F, message=F}
suppressMessages(library("dplyr"))
suppressMessages(library("ggplot2"))
suppressMessages(library("susieR"))
suppressMessages(library("broom"))
```

# Introduction

In this tutorial, we are going explore explore how association testing works. Our trait of interest is the expression level of the CTLA4 gene in CD4+ T-cells in the OneK1K dataset and we are going to explore how genetic variants near that gene are associated with its expression level. 

## Importing the data

All of the data that you need for this tutorial is stored in the OneK1K_CTLA4_data.tsv.gz file. Let's first read it in

```{r}
ctla4_df = readr::read_tsv("OneK1K_CTLA4_data.tsv.gz", show_col_types = FALSE)
ctla4_df
```
As you can see, the dataset has 981 rows corresponding to 981 individuals. The first three four columns contain some metadata about the sample, the fifth columm (ctla4) contain the normalised expression level of the CTLA4 gene in each sample. The values that you can see here are after inverse normal transformation (INT) which means that the gene expression values across individuals are normally distributed with zero mean and unit standard deviation. This is done to avoid potential outlier values that could bias linear regression.

You can check that easily:
```{r}
hist(ctla4_df$ctla4)
```

The remaining columns correspond to 1390 genetic variants from this gene region. Notice that even in this relatively narrow region of the genome (~400 kb or 400,000 bp) the number of genetic variants (p = 1390) already exceeds the number of samples (n = 981).

## Association testing

As mentioned in the [video lecture](https://www.youtube.com/watch?v=pglYf7wocSI), association testing is typically performed one genetic variant at a time using a simple linear regression. For example to test for an association between the chr2_203874196_G_A genetic variant and CTLA4 expression, we could use this R command:

```{r}
model = lm(ctla4 ~ chr2_203874196_G_A, ctla4_df)
model
```
We can also easily add in additional covariates such as the biological sex of the sample:

```{r}
model = lm(ctla4 ~ chr2_203874196_G_A + sex, ctla4_df)
model
```

Finally, we can use the broom package to extract effect sizes, standard errors and p-values from the model as a nicely formatted table:

```{r}
res = broom::tidy(model)
res
```

As you can see, there is very strong association between the genotype of the chr2_203874196_G_A variant and CTLA4 expression (effect size = 0.65, p-value = 7.53e-43)

We can also visualise this with a boxplot:

```{r}
ggplot(ctla4_df, aes(x = as.factor(round(chr2_203874196_G_A)), y = ctla4)) + 
  geom_boxplot() + 
  geom_jitter(width = 0.2)
```

Now, to get the same results for all 1390 genetic variants in the region, we could run the lm() command 1390 times, but this would be quite annoying (and very slow). Fortunately there are a number of software packages such as MatrixEQTL (R), TensorQTL (Python) and FastQTL (command-line) to run these computations efficiently as matrix operations. Since we eventually, want to perform fine mapping, we are going to use the susieR R package to also get the single-variant linear regression results.

#Preparing data for fine mapping

The SuSiE method requires two inputs: (1) vector of trait values (CTLA4 expression in our case) and (2) a matrix of standardised genotypes (where all varaints have zero mean across individuals). Let's standardise the genotypes first:

```{r}
gt_matrix = t(as.matrix(ctla4_df[,-c(1:5)]))
standard_genotypes = t(gt_matrix - apply(gt_matrix, 1, mean))
standard_genotypes[1:5,1:5]
```
Now, let's run SuSiE:

```{r}
fitted <- susieR::susie(standard_genotypes, ctla4_df$ctla4,
                        verbose = TRUE,
                        compute_univariate_zscore = TRUE,
                        coverage = 0.95)
fitted$variant_id = colnames(standard_genotypes)
```
and extract the univariate z scores:
```{r}
z_df = dplyr::tibble(variant = fitted$variant_id, z = fitted$z) %>%
  dplyr::mutate(variant_rank = c(1:n()))
z_df
```

We can also visualise the association signal:
```{r}
ggplot(z_df, aes(x = variant_rank, y = abs(z))) + geom_point()
```

Calculate LD between the lead variant (=variant with the largest absolute Z score) and all other variants:

```{r}
ld_matrix = cor(standard_genotypes)^2
ld_with_lead = ld_matrix["chr2_203874196_G_A",]
ld_df = dplyr::tibble(ld = ld_with_lead, variant = names(ld_with_lead))
ld_df
```
Visualise the relationship between LD with the lead variant and Z score:
```{r}
joined_df = dplyr::left_join(ld_df, z_df, by = "variant")
ggplot(joined_df, aes(x = abs(z), y = ld)) + geom_point()
```

As you can see, there are many variants with large absolute Z scores, but their Z scores are almost perfectly explained by their LD (r2) with the lead variant. 

# Fine mapping

Visualise fine mapping results on top of -log10 p-values z-scores from the univariate analysis:

```{r}
susieR::susie_plot(fitted, y = "z")
```

Visualise posterior inclusion probabilities (PIP) for each variant directly

```{r}
susieR::susie_plot(fitted, y = "PIP")
```

Extract variant belonging to the credible set

```{r}
credible_set = fitted$pip[fitted$sets$cs$L1]
names(credible_set) = fitted$variant_id[fitted$sets$cs$L1]
credible_set
```