# Assignment 7

## Task 1: Variance explained and statistical power to detect genetic associations (2 points).
**Learning objective**: Understand the relationships between genotype allele frequency, variance explained, sample size and the statistical power to detect genetic associations.

To help you get started with this exercise, I have created a small [example](https://github.com/kauralasoo/MTAT.03.239_Bioinformatics/blob/master/QTL_analysis/simulating_genetic_associations.md) in R to illustrate how we can simulate the effect of genetic variants on gene expression. In your solution you can use either my code snippets or equivalent functions in Python. 

Statistical power of an hypothesis test is the probability that the test rejects the null hypothesis (H0) when a specific alternative hypothesis (H1) is true ([Wikipedia](https://en.wikipedia.org/wiki/Power_%28statistics%29)). 

Please answer the following questions. 

1. Using the code example provided [here](https://github.com/kauralasoo/MTAT.03.239_Bioinformatics/blob/master/QTL_analysis/simulating_genetic_associations.md), estimate the number of samples needed so that the likelihood ratio test (LRT) has 80% power to reject the null hypothesis (p < 0.05) that genotype has no effect on gene expression. Assume that the minor allele frequency (MAF) of the genetic variant is 0.5 and genotype explains 50% of the variance in the gene expression data (the remaining 50% of the variance is due to measurement noise). You can estimate power at a given sample size by generating many (>100) simulated datasets, performing the likelihood ratio test and counting how often is the p-value less then the desired threshold (p<0.05).
2. Modify the simulation so that the minor allele frequency is still 0.5, but the genotype now explains only 10% of the total variance. How much larger has the sample size to be to achieve the same statistical power of 80%?
3. Start with the original simulation in point 1, but reduce the minor allele frequency to 10%. How does this influence the fraction of variance explained by the genotype when the measurement noise remains unchanged? How large sample size will you need to achieve 80% power?

## Task 2: Understanding correlation (linkage disequilibrium (LD)) between genetic variants (1 point)
High correlation (linkage disequilibrium (LD)) between genetic variants means that it is challenging to identify which of the many associated variants is the causal variant. One way to quantify LD between two genetic variants is to calculate the square of the Pearson's correlation coefficient (r<sup>2</sup>). Based on the genotype data presented in the [flow cytometry tutorial](https://github.com/kauralasoo/flow_cytomtery_genetics/blob/master/analysis/variance_components/estimate_variance_components.md),  estimate the number of genetic variants that are in high LD (r<sup>2</sup> > 0.8) with rs778587 genetic variant. In this example, rs778587 is the genetic variants that is most strongly associated with the expression of the CD14 protein. The variant with the strongest association in a region is often referred to as the *lead* variant.

The original VCF file is available from [here](https://github.com/kauralasoo/flow_cytomtery_genetics/blob/master/data/genotypes/open_access_genotypes.vcf.gz)  and an R data file containing the genotype matrix where the genotypes have been converted to allele counts (0,1,2) is available from [here](https://github.com/kauralasoo/flow_cytomtery_genetics/blob/master/data/genotypes/open_access_genotypes.rds). You can open the .rds file in R using the `readRDS()` command.  In your report, include the code that you used to calculate LD as well es the number of genetic variants with r<sup>2</sup> > 0.8 with rs778587. Remember that LD is the correlation between two genetic variants across individuals.

Note that the flow cytometry data is in its own GitHub repository: https://github.com/kauralasoo/flow_cytomtery_genetics.

## Task 3. Understanding the basics of fine mapping (1.5 points)
Using the flow cytometry fine mapping [tutorial](https://github.com/kauralasoo/flow_cytomtery_genetics/blob/master/analysis/fine_mapping/fine_mapping_example.md) and gene expression data from the naive macrophages, perform fine mapping on the CD14 gene expression data. You can downoload the  normalised expression data that from [Zenodo](https://zenodo.org/record/2571453#.XpSCQFMza-4) (salmonella_cqn_matrix.txt.gz file). You basically just need to replace the protein fluorescent intensity values used in the tutorial with normalised gene expression values for CD14 gene that you can download from Zenodo. Compare the the 95% credible set to the one obtained from protein expression (fuorescent intensity) data in this tutorial. 
1. How big are to two credible sets? How many variants are present in both credible sets? 
2. Do you think it is likely that the CD14 gene and protein expression are regulated by the same causal genetic varaint? Why or why not? How strong to you think the fine mapping evidence is? What other evidence can you take into account (think [Assignment 6](https://github.com/kauralasoo/MTAT.03.239_Bioinformatics/blob/master/assignments/2024/Assignment_6.md), Task 2).

## Task 4. Exploring the ARHGEF3 locus (1 points)
For this task, first download the CEDAR gene expression and genotype data from [here](https://drive.google.com/drive/folders/1PveWl7nJQlyYp_n2kY79qOx4LrwS9GcV?usp=sharing). Next, follow the [ARHGEF3 tutorial](https://github.com/kauralasoo/MTAT.03.239_Bioinformatics/blob/master/fine_mapping/CEDAR_finemapping_example.md) to get a basic working example. 

1. Look up the ARHGEF3 fine-mapped variant from the [Open Targets Genetics Portal](https://genetics.opentargets.org/). In which cell types and tissues can you see the eQTL? What are other traits most strongly associated with this variant? Do these signals also colocalise with the ARHGEF3 eQTL? How strong is the evidence for colocalisation? (Hint: this example was shown in the lecture). You might also find the [documentation](https://genetics-docs.opentargets.org/our-approach/colocalisation-analysis) for the Open Targets Genetics portal useful. The eQTL data in the Open Targets Genetics Portal comes from the [eQTL Catalogue](https://www.biorxiv.org/content/10.1101/2020.01.29.924266v1) that we have developed here in Tartu together with our collegues at the European Bioinformatics Institute.

## Task 5. Perform fine mapping for the lactase (LCT) gene in the illeum tissue (2.5 points)
In Task 1 of [Assignment 5](https://github.com/kauralasoo/MTAT.03.239_Bioinformatics/blob/master/homeworks/2024/Assignment_5.md) you compared the frequenies of [rs4988235](https://www.snpedia.com/index.php/Rs4988235) genetic variant accross populations. As mentioned in Homework 4, rs4988235 is associated with the ability to drink milk in Europeans populations and is thought to act by increasing the expression of the LCT gene in the correct tissue. Modify the fine mapping example in the ARHGEF3 tutorial to perform fine mapping for the lactase (LCT) gene in the [ileum](https://en.wikipedia.org/wiki/Ileum) tissue (which should probably be close to the correct tissue in which to study LCT expression). Note that since the LCT gene has two probes on the microarray used by the CEDAR study, you should perform fine mapping separately for both probes (Hint: for one probe the analysis will probably fail and you will not find any credible sets).
1. How many variants are part of the 95% credible set? What is the the highest PIP value observed across all variants?
2. Is the rs4988235 variant part of the credible set?
3. Increase the credible set coverage to 99%. How many variants are now in the credible set? Is rs4988235 one of them?
4. Based on these results, can we concluded that the LCT eQTL observed in ileum tissue is responsible for the lactose persistence phenotype? 
5. Open Targets Genetics Portal has GWAS results for the UK Biobank questionaire item ["Never/rarely have milk | milk type used"](https://genetics.opentargets.org/study/NEALE2_1418_6) that has a significant association near the LCT gene. Which eQTLs (genes) does this association colocalise with? In which cell types and tissues? Is LCT one of them?

